src/spectralanalysis.jl

```diff
@@ -1,7 +1,27 @@
 using Arpack
+import KrylovKit: eigsolve
 
 const nonhermitian_warning = "The given operator is not hermitian. If this is due to a numerical error make the operator hermitian first by calculating (x+dagger(x))/2 first."
 
+mutable struct DiagStrategy{DT}
+    n::Int
+    v0::Union{Nothing, AbstractVector}
+end
+DiagStrategy{T}(n::Int) where T = DiagStrategy{T}(n, nothing)
+const LapackDiag = DiagStrategy{:lapack}
+const ArpackDiag = DiagStrategy{:arpack}
+const KrylovDiag = DiagStrategy{:krylov}
+
+arithmetic_unary_error = QuantumOpticsBase.arithmetic_unary_error
+DiagStrategy(op::AbstractOperator) = arithmetic_unary_error("DiagStrategy", op)
+function DiagStrategy(op::DataOperator)
+    basis(op) # Checks basis match
+    DiagStrategy(op.data)
+end
+DiagStrategy(m::Matrix) = LapackDiag(size(m)[1])
+DiagStrategy(::SparseMatrixCSC) = KrylovDiag(6)
+DiagStrategy(m::AbstractMatrix) = ArgumentError("Cannot detect DiagStrategy for array type $(typeof(m))")
```

Comment:
Could you elaborate on this error message. For the last year or more SciML has undertaken a really great process of writing as helpful as possible error messages. For instance, could you add something along the lines of `This error probably is raised because you have tried to ... To fix it, consider doing ...`